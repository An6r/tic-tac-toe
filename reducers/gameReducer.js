import _ from 'lodash';
import * as actionTypes from '../actions/actionTypes';
import {
  BOARD_SIZE,
  generateBoardGrid,
  verifyBoardCells
} from '../helpers/game';

export const initialState = {
  boardGrid: generateBoardGrid(),
  currentPlayer: 1,
  playersSymbols: {
	  1: 'X',
	  2: 'O'
  },
  winnerMarks: {},
  winnerPlayer: null,
  marksCount: 0,
  isTie: false,
};

export const gameReducer = (state = initialState, action) => {
  switch (action.type) {
    case actionTypes.NEW_SIGN: {
      const { currentPlayer } = state;
      const boardGrid = _.cloneDeep(state.boardGrid); 

      const { column, row } = action;

      boardGrid[column][row] = currentPlayer;

      return {
        ...state,
        boardGrid,
        marksCount: state.marksCount + 1,
      };
    }
    case actionTypes.VERIFY_BOARD: {
      const boardSize = BOARD_SIZE;
      const maxMarksCount = boardSize ** 2;

      const {
		  currentPlayer: winnerPlayer,
		  marksCount
	  } = state;

      if (marksCount === maxMarksCount) {
        return {
          ...state,
          isTie: true,
        };
      }

      const boardGrid = _.cloneDeep(state.boardGrid);

      for (let i = 0; i < boardSize; i += 1) {
        // Check if user won horizontally
        if (verifyBoardCells(boardGrid[i][0], boardGrid[i][1], boardGrid[i][2])) {
          return {
            ...state,
            winnerMarks: {
              [i]: [0, 1, 2],
            },
            winnerPlayer,
          };
        }

        // Check if user won vertically
        if (verifyBoardCells(boardGrid[0][i], boardGrid[1][i], boardGrid[2][i])) {
          return {
            ...state,
            winnerMarks: {
              0: [i],
              1: [i],
              2: [i],
            },
            winnerPlayer,
          };
        }
      }

      /*
       Check if user won diagonally
       */
      if (verifyBoardCells(boardGrid[0][0], boardGrid[1][1], boardGrid[2][2])) {
        return {
          ...state,
          winnerMarks: {
            0: [0],
            1: [1],
            2: [2],
          },
          winnerPlayer,
        };
      }

      if (verifyBoardCells(boardGrid[0][2], boardGrid[1][1], boardGrid[2][0])) {
        return {
          ...state,
          winnerMarks: {
            0: [2],
            1: [1],
            2: [0],
          },
          winnerPlayer,
        };
      }

      return state;
    }
    case actionTypes.NEXT_PLAYER: {
      return {
        ...state,
        currentPlayer: ((state.currentPlayer === 1) ? 2 : 1),
      };
    }
    case actionTypes.RESTART_GAME: {
      return {
        ...initialState
      };
    }
    default: {
      return state;
    }
  }
};