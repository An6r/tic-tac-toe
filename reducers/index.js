import { gameReducer as game } from './gameReducer';

export default {
	game
};
