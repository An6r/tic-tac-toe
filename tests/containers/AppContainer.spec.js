import React from 'react';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import { initialState as game } from '../../reducers/gameReducer';
import ConnectedApp, { AppContainer } from '../../containers/AppContainer';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

function setupComponent() {
  const props = {
    ...game,
    actions: {
      restartGame: jest.fn()
    },
  };

  const wrapper = shallow(
    <AppContainer {...props} />,
  );

  return {
    props,
    wrapper
  };
}

describe('<App />', () => {
  it('should render without exploding', () => {
    const store = mockStore({
      game,
    });

    const wrapper = shallow(
      <ConnectedApp store={store} />,
    );

    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('should restart the game on button click', () => {
    const { wrapper, props } = setupComponent();

    wrapper.find('Button[name="restart"]').simulate('click');

    expect(props.actions.restartGame.mock.calls.length).toBe(1);
  });
});
