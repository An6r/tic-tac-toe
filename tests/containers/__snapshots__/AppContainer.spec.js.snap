// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should render without exploding 1`] = `
<AppContainer
  actions={
    Object {
      "newSignOnBoard": [Function],
      "nextPlayer": [Function],
      "restartGame": [Function],
      "verifyBoard": [Function],
    }
  }
  boardGrid={
    Array [
      Array [
        null,
        null,
        null,
      ],
      Array [
        null,
        null,
        null,
      ],
      Array [
        null,
        null,
        null,
      ],
    ]
  }
  currentPlayer={1}
  isTie={false}
  marksCount={0}
  playersSymbols={
    Object {
      "1": "X",
      "2": "O",
    }
  }
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  winnerMarks={Object {}}
  winnerPlayer={null}
/>
`;
